src/wallet/mod.rs

```diff
@@ -227,112 +224,82 @@ where
-    /// check is necessary for backwards compatibility of a checksum-inception bug.
-    fn db_checksum(db: &mut D, desc: &str, kind: KeychainKind) -> Result<(), Error> {
-        let checksum = calc_checksum_bytes_internal(desc, true)?;
-        if db.check_descriptor_checksum(kind, checksum).is_ok() {
-            return Ok(());
-        }
-
-        let checksum_inception = calc_checksum_bytes_internal(desc, false)?;
-        db.check_descriptor_checksum(kind, checksum_inception)
-    }
-
     /// Get the Bitcoin network the wallet is using.
     pub fn network(&self) -> Network {
         self.network
     }
 
-    // Return a newly derived address for the specified `keychain`.
-    fn get_new_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
-        let incremented_index = self.fetch_and_increment_index(keychain)?;
+    /// Iterator over all keychains in this wallet
+    pub fn keychanins(&self) -> impl ExactSizeIterator<Item = KeychainKind> {
+        let mut keychains = vec![KeychainKind::External];
 
-        let address_result = self
-            .get_descriptor_for_keychain(keychain)
-            .at_derivation_index(incremented_index)
-            .address(self.network);
+        if self.change_descriptor.is_some() {
+            keychains.push(KeychainKind::Internal);
+        }
 
-        address_result
-            .map(|address| AddressInfo {
-                address,
-                index: incremented_index,
-                keychain,
-            })
-            .map_err(|_| Error::ScriptDoesntHaveAddressForm)
+        keychains.into_iter()
     }
 
-    // Return the the last previously derived address for `keychain` if it has not been used in a
-    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
-    fn get_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
-        let current_index = self.fetch_index(keychain)?;
+    // Return a newly derived address for the specified `keychain`.
+    fn get_new_address(&mut self, keychain: KeychainKind) -> AddressInfo {
+        let next_index = self.next_derivation_index(keychain);
 
-        let derived_key = self
+        let definite_descriptor = self
             .get_descriptor_for_keychain(keychain)
-            .at_derivation_index(current_index);
+            .at_derivation_index(next_index);
 
-        let script_pubkey = derived_key.script_pubkey();
+        self.spk_tracker
+            .add_spk((keychain, next_index), definite_descriptor.script_pubkey());
 
-        let found_used = self
-            .list_transactions(true)?
-            .iter()
-            .flat_map(|tx_details| tx_details.transaction.as_ref())
-            .flat_map(|tx| tx.output.iter())
-            .any(|o| o.script_pubkey == script_pubkey);
+        let address = definite_descriptor
+            .address(self.network)
+            .expect("descriptor must have address form");
```

Comment:
Does this mean that we're dropping support for descriptors without an address form, or that we trust our users to not call `get_new_address` if their descriptor can't have an address form?

Reply:
The latter I think. Needs documentation.